### PR TITLE
KAFKA-9011: Scala bindings for flatTransform and flatTransformValues in KStream

### PR DESCRIPTION
**Pull request by kokachev**
@cadonna 

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @bbejeck @vvcephei

**Comment by vvcephei** [comment]
Hey @kokachev ,

Thanks for the contribution! This looks good to me, but can you add some tests exercising the new methods? We have had quite a few issues in the past with Scala implicits behaving unintuitively, so now it's hard to trust even code that looks "obviously correct".

Thanks,
-John

**Review by bbejeck — COMMENTED** [review]
Thanks @kokachev over LGTM, but I agree with @vvcephei that we'll need some tests using the new methods.

**Comment by kokachev** [comment]
@bbejeck , @vvcephei, Added tests for new functions as requested. Got a bit excited and added test for transform, which wasn't in a scope  ;)

**Comment by kokachev** [comment]
retest this please

**Comment by kokachev** [comment]
retest this please

**Comment by kokachev** [comment]
@bbejeck , @vvcephei , @mjsax Can you guys please merge it to  2.3.2 or 2.4? JDK8 build is failing due to test broken upstream.

**Comment by kokachev** [comment]
retest this please

**Review by vvcephei — APPROVED** [review]
Hey @kokachev ,

Sorry for the delay. These tests look great to me. I think we can go ahead and merge this PR!

Thanks,
-John

**Comment by vvcephei** [comment]
(doing it again so we can capture the test results)

Retest this, please.

**Comment by vvcephei** [comment]
Note that it's too late for the 2.4.0 release, as of the beginning of October, so I think we need to wait for the 2.4.0 release to actually happen before cherry-picking to the 2.4 branch.

We could cherry-pick it to the 2.3 branch, but it might be safer to do the 2.4 and 2.3 cherry-picks at the same time (so we don't drop 2.4 on the floor).

**Comment by kokachev** [comment]
@vvcephei ,
Thanks for getting back on this.
I think it's relatively low risk patch, because it adds new functions without changing existing code. It fixes a "bug" of Scala API being unjustly treated as inferior compared to  Java API :-)  The only thing that can potentially fail is tests.

**Review by bbejeck — APPROVED** [review]

**Comment by bbejeck** [comment]
Merged #7520 into trunk.

**Comment by bbejeck** [comment]
Thanks for the contribution @kokachev!

**Comment by bbejeck** [comment]
FWIW I agree with @vvcephei with the cherry-picking, IMHO it's better to wait and do them at one time.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/FunctionsCompatConversions.scala [hunk 1]
```diff
@@ -120,4 +122,40 @@ private[scala] object FunctionsCompatConversions {
...
   }
+
+  implicit class TransformerSupplierAsJava[K, V, VO](val supplier: TransformerSupplier[K, V, Iterable[VO]])
+      extends AnyVal {
+    def asJava: TransformerSupplier[K, V, JIterable[VO]] = new TransformerSupplier[K, V, JIterable[VO]] {
+      override def get(): Transformer[K, V, JIterable[VO]] =
+        new Transformer[K, V, JIterable[VO]] {
+          override def transform(key: K, value: V): JIterable[VO] = supplier.get().transform(key, value).asJava
+          override def init(context: ProcessorContext): Unit = supplier.get().init(context)
+          override def close(): Unit = supplier.get().close()
```
Comment (bbejeck) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/FunctionsCompatConversions.scala:133]: @kokachev I just realized that we have multiple calls to `supplier.get()` here.  Calling `supplier.get()` will return a new instance each time correct?  This needs to get changed here and below, as well.  We can do this in a follow-up PR using KAFKA-9011.

**Pull request merged**